File: solutions/111.minimum-depth-of-binary-tree/minimum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if(!root) return 0;
        int left = minDepth(root -> left);
        int right = minDepth(root -> right);
        if(root -> left) {
            if(root -> right) return min(left, right) + 1;
            return left + 1;
        } else if(root -> right) return right + 1;
        return 1;
    }
};

File: solutions/111.minimum-depth-of-binary-tree/minimum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int depth = 1;
        TreeNode curNode = null;
        while (!queue.isEmpty()) {
            int size = queue.size();
            while(size-- > 0) {
                curNode = queue.remove();
                if (curNode.left == null && curNode.right == null) return depth;
                if (curNode.left != null) queue.offer(curNode.left);
                if (curNode.right != null) queue.offer(curNode.right);
            }
            depth++;
        }
        return depth;
    }
}